src/plugins/intel_cpu/src/nodes/transpose.cpp

```diff
@@ -76,9 +77,8 @@ class TransposeShapeInferFactory : public ShapeInferFactory {
         if (const auto order = ov::as_type_ptr<const ov::op::v0::Constant>(m_op->get_input_node_shared_ptr(ov::op::v1::Transpose::ORDER))) {
             const auto axes_vec = order->cast_vector<size_t>();
             return std::make_shared<TransposeShapeInfer>(m_op->get_output_partial_shape(0).rank().get_length(), axes_vec);
-        } else {
-            IE_THROW() << "The second input is NOT a constant node, which is unexpected in the Transpose shape inference factory";
         }
+        return std::make_shared<TransposeShapeInfer>();
```

Comment:
The main problem is that the node instantiation pipeline try to instantiate a node of a specific type even though the specific configuration is not supported by the plugin (isSupportedOperation returns false). That means that the base `Node` class constructor is always called, and the factory also called, notwithstanding the fact that the node object will not be created due to exception in the specific type constructor. Thus, we have two options here:
1. The factory may return NgraphShapeInfer object in case the specific configuration is not supported. That is we should store the ngraph op in the factory and in the else branch instead of throwing exception we should instantiate NgraphShapeInfer factory for this op with corresponding port mask and call `makeShapeInfer()` to that factory object. That will automatically make the support for a non constant `ORDER` input working, however the `ngraph` implementation is suboptimal.
2. We can make yet another class `TransposeDynInputShapeInfer` which should support non constant data on the `ORDER` input, but its implementation will throw `not implemented` exception since at the current moment of time the CPU plugin `Transpose` node does not support such configuration and when someone removes this constraint he will be forced to update the specific shape inference implementation as well. So the factory will simply return an instance of that class on the else branch.
Personally I prefer the second option, since it is not silently work with bad performance.

Reply:
The second option sounds good and is friendly.
I update a TransposeDynShapeInfer class for the UnConstant Order. In such class, the constructor is empty and the infer function throws *NotImplemented* info.